sound/soc/sof/intel/hda-loader-skl.c

```diff
@@ -0,0 +1,520 @@
+{
+	cl_skl_cldma_cleanup_spb(sdev);
+	cl_skl_cldma_stream_clear(sdev);
+	snd_dma_free_pages(&sdev->dmab);
+	snd_dma_free_pages(&sdev->dmab_bdl);
+	sdev->dmab.area = NULL;
+}
+
+static int cl_dsp_init_skl(struct snd_sof_dev *sdev)
+{
+	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
+	const struct sof_intel_dsp_desc *chip = hda->desc;
+	unsigned int sts;
+	int ret;
+
+	/*
+	 * check if the core is already enabled, if yes, reset and
+	 * make it run. If not, powerdown and enable it again.
+	 */
+	if (hda_dsp_core_is_enabled(sdev, HDA_DSP_CORE_MASK(0))) {
+
+		/* if enabled, reset it, and run the core. */
+		ret = hda_dsp_core_stall_reset(sdev, HDA_DSP_CORE_MASK(0));
+		if (ret < 0)
+			goto err;
+
+		ret = hda_dsp_core_run(sdev, HDA_DSP_CORE_MASK(0));
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: dsp core start failed %d\n",
+				ret);
+			goto err;
+		}
+	} else {
+		/*
+		 * if not enabled, power down it first and
+		 * then powerup and run the core.
+		 */
+		ret = hda_dsp_core_reset_power_down(sdev, HDA_DSP_CORE_MASK(0));
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: dsp core0 disable fail: %d\n", ret);
+			goto err;
+		}
+		ret = hda_dsp_enable_core(sdev, HDA_DSP_CORE_MASK(0));
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: dsp core0 enable fail: %d\n", ret);
+			goto err;
+		}
+	}
+
+	/* prepare DMA for code loader stream */
+	ret = cl_stream_prepare_skl(sdev);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: dma prepare fw loading err: %x\n",
+			ret);
+		return ret;
+	}
+
+	/* enable IPC interrupts */
+	hda_dsp_ipc_int_enable(sdev);
+
+	/* polling the ROM init status information. */
+	ret = snd_sof_dsp_read_poll_timeout(sdev, HDA_DSP_BAR,
+					    HDA_ADSP_FW_STATUS_SKL, sts,
+					    ((sts & SKL_DSP_ROM_SKL_STS_MASK)
+						    == SKL_DSP_ROM_INIT_DONE),
+					    HDA_DSP_REG_POLL_INTERVAL_US,
+					    chip->rom_init_timeout *
+					    USEC_PER_MSEC);
+	if (ret < 0)
+		goto err;
+
+	return ret;
+
+err:
+	hda_dsp_dump_skl(sdev, SOF_DBG_REGS | SOF_DBG_PCI | SOF_DBG_MBOX);
+	cl_cleanup_skl(sdev);
+	hda_dsp_core_reset_power_down(sdev, HDA_DSP_CORE_MASK(0));
+	return ret;
+}
+
+static void cl_skl_cldma_fill_buffer(struct snd_sof_dev *sdev,
+				     unsigned int size,
+				     const void *curr_pos,
+				     bool intr_enable, bool trigger)
+{
+	/* 1. copy the image into the buffer. */
+	memcpy(sdev->dmab.area, curr_pos, size);
+
+	/* 2. Set the interrupt. */
+	if (intr_enable)
+		cl_skl_cldma_set_intr(sdev, true);
+
+	/* 3. Set the SPB. */
+	cl_skl_cldma_setup_spb(sdev, size, trigger);
+
+	/* 4. Trigger the code loading stream. */
+	if (trigger)
+		cl_skl_cldma_stream_run(sdev, true);
+}
+
+static int
+cl_skl_cldma_copy_to_buf(struct snd_sof_dev *sdev, const void *bin,
+			 u32 total_size, u32 bufsize)
+{
+	unsigned int bytes_left = total_size;
+	const u8 *curr_pos = (u8 *)bin;
+
+	if (total_size <= 0)
+		return -EINVAL;
+
+	while (bytes_left > 0) {
+
+		if (bytes_left > bufsize) {
```

Comment:
this could be reverted like
```
	if (bytes_left <= bufsize) {
		...
		return 0;
	}
```
to avoid the additional indentation level. caused by the `else`